keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,854 @@
+
+## Motivation
+
+**Reliability:** Admission webhooks continue to be an operational sore spot for many Kubernetes
+users. Webhooks that target cluster critical resources put the admission controller backing the
+webhook in the critical path of cluster stability. Even if tools like namespace scoping are used to
+avoid circular-dependencies and exclude critical system resources, a webhook outage can still have a
+major impact on cluster availability. This proposal aims to mitigate (but not eliminate) these
+issues by allowing webhooks to be more narrowly scoped and targeted.
+
+
+The same benefits also apply to bootstrapping a new cluster, or in a disaster-recovery scenario
+where objects in a cluster need to be recreated from source code. Being able to define and deploy a
+webhook before any workload Pods are defined means that admins don't need to bypass security
+controls to fix things, and helps to ensure there's never a security gap when restoring service.
+
+**Performance:** Admission webhooks sit in the critical request path for write-requests. Validating
+webhooks can be run in parallel, but Mutating webhooks must be run in serial (up to 2 times!). This
+makes webhooks extremely latency sensitive, and even a webhook that doesn't do any work still needs
+to pay the network round-trip cost.
+
+**Supportability:** For hosted or managed Kubernetes distributions, webhooks can be a problem when
+they interfere with requests by managed components. The existing criteria for filtering out requests
+are insufficient for many use cases, and aren't easily appended with provider rules.
+
+_What about [CEL for Admission Control](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control)?_
+`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
+for admission webhooks, but it is intentionally not attempting to cover every possible use case.
+This proposal aims to improve the situation for those webhooks that cannot be migrated.
+
+### User Stories
+
+#### Exclude resources from a wildcard rule
+
+> I want to enforce metadata policy through an admission webhook without adding latency & risk to
+> high QPS system requests.
+
+Currently, if a webhook uses wildcard match rules, there is no way to filter out a subset of
+resources or requests from matching the wildcard. If the webhook instead enumerates every resource
+that should match, it must be kept up-to-date with every CRD that's added.
+
+With CEL match conditions, the webhook could specify wildcard match rules, and add match conditions
+to filter out the desired resources:
+
+```yaml
+rules:
+  # Match CREATE & UPDATE on all resources:
+  - operations:
+    - CREATE
+    - UPDATE
+    apiGroups: '*'
+    apiVersions: '*'
+    resources: '*'
+matchConditions:
+  # Exclude leases from the webhook
+  - expression: '!(request.resource.group == "coordination.k8s.io" && resource.resource == "leases")'
```

Comment:
I suggest you make these {name, expression} pairs. That way it would be easy to have a controller own and maintain one or more named expressions. The name helps explain what the expression is for, and it make SSA work well. The name could also be reported/logged in errors etc.